src/plugins/discover/public/application/context/context_app.tsx

```diff
@@ -140,6 +177,61 @@ export const ContextApp = ({ dataView, anchorId }: ContextAppProps) => {
     [filterManager, dataViews, dataView, capabilities]
   );
 
+  const updateAdHocDataViewId = useCallback(
+    async (dataViewToUpdate: DataView) => {
+      const updatedDataView = await dataViews.create({
+        ...dataViewToUpdate.toSpec(),
+        id: undefined,
+      });
+      dataViews.clearInstanceCache(dataViewToUpdate.id);
+      return updatedDataView;
+    },
+    [dataViews]
+  );
+
+  const onAdHocDataViewChange = useCallback(async () => {
+    const updatedDataView = await updateAdHocDataViewId(dataView);
+
+    const uiActions = await getUiActions();
+    const trigger = uiActions.getTrigger(UPDATE_FILTER_REFERENCES_TRIGGER);
+    const action = uiActions.getAction(UPDATE_FILTER_REFERENCES_ACTION);
+
+    action?.execute({
+      trigger,
+      fromDataView: dataView.id,
+      toDataView: updatedDataView.id,
+      usedDataViews: [],
+    } as ActionExecutionContext);
+
+    // get just updated filters, line order is important here
+    const updatedFilters = [...filterManager.getGlobalFilters(), ...filterManager.getAppFilters()];
+    replaceContextLocation(services.contextLocator, {
+      rowId: anchorId,
+      dataViewSpec: updatedDataView.toSpec(false),
+      columns,
+      filters: updatedFilters,
+      timeRange: locationState.timeRange,
+      query: locationState.query,
+    });
+  }, [
+    anchorId,
+    columns,
+    dataView,
+    filterManager,
+    locationState.query,
+    locationState.timeRange,
+    services.contextLocator,
+    updateAdHocDataViewId,
+  ]);
+
+  const onFieldEdited = useCallback(async () => {
+    if (!dataView.isPersisted()) {
+      onAdHocDataViewChange();
```

Comment:
Wow, there's a lot of improvement in this PR 👍 
Here's the part that I'm thinking about. In the context / view surrounding documents view, we enable the user to edit field settings just in the table header
![Discover_-_Elastic](https://user-images.githubusercontent.com/463851/195572180-5d4ae03c-11c8-415d-9da3-1e8d206b6d32.png)
This adds complexity to the code to support ad hoc data views, and I'm thinking, should we even allow editing of fields in the surrounding view? functionality wise it's limited already (no field selector, no data view selection, no sharing, ...) 

I could imagine that this view might also become a building block, so it could be a flyout of the main view, focused on the primary task ... to view document in the context

I'm interested in the opinion of @elastic/kibana-data-discovery
Should or shouldn't we keep the "Edit data view field" functionality in the context view?

Reply:
IMO it would be fine to remove it if it adds significant complexity. Like you said, most other data view features are missing from surrounding docs already since it serves a pretty specific purpose. I doubt many (any?) would miss it here personally.

Reply:
@kertal From the UX perspective, then the user will treat this view as a "normal" Discover table view where he/she/they have the possibility to **edit data view field**.  When we start working on making surrounding documents as a building block and perhaps within the flyover, then we can start changing things. So I would park this until we reach that point. We have other more prioritized "building blocks" that need to be finished. 

I am voting for keeping it there, since from the user perspective they are using a table with a filter "surrounding documents" and visit it again later.

Reply:
> I am voting for keeping it there, since from the user perspective they are using a table with a filter "surrounding documents" and visit it again later.

I would agree if this was a local state the user would be editing. But in this case, it's actually the data view where the modification is persisted. In the Discover main view, we have all elements in place allowing the modification of the data view, switching of data view, users can add columns to the table, add runtime fields etc. While this was a little change initially, the recent modification to support editing of ad-hoc ones, is more complex. I think it's not obvious to the user what happens here. I think we have 2 possible paths how we could develop this part of Discover
**a)** Enabling full support of editing functionality 
**b)** No editing functionality is provided, just local changes 
For now I think **b** is the better way, and when we create also building block, we can revisit it and discuss it

Reply:
@kertal - Let's go with your b) with a comment that since we are making this a "view" mode, then we need to change it from looking the same as "edit" mode in Discover. Let's put a task for surrounding documents "view" mode in the backlog.

Reply:
Done.